pkg/mutation/match.go

```diff
@@ -0,0 +1,136 @@
+package mutation
+
+import (
+	"fmt"
+
+	mutationsv1 "github.com/open-policy-agent/gatekeeper/apis/mutations/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// Matches verifies if the given object belonging to the given namespace
+// matches the current mutator.
+func Matches(match mutationsv1.Match, obj *unstructured.Unstructured, ns *corev1.Namespace) (bool, error) {
+	meta, err := meta.Accessor(obj)
+	if err != nil {
+		return false, fmt.Errorf("Accessor failed for %s", obj.GetObjectKind().GroupVersionKind().Kind)
+	}
+
+	foundMatch := false
+
+	for _, kk := range match.Kinds {
+		kindMatches := false
+		groupMatches := false
+
+		for _, k := range kk.Kinds {
+			if k == "*" || k == obj.GroupVersionKind().Kind {
+				kindMatches = true
+				break
+			}
+		}
+		for _, g := range kk.APIGroups {
+			if g == "*" || g == obj.GroupVersionKind().Group {
+				groupMatches = true
+				break
+			}
+		}
+		if kindMatches && groupMatches {
+			foundMatch = true
+		}
+	}
+	if len(match.Kinds) == 0 {
+		foundMatch = true
+	}
+
+	if !foundMatch {
+		return false, nil
+	}
+
+	if match.Scope == apiextensionsv1beta1.ClusterScoped &&
+		meta.GetNamespace() != "" {
+		return false, nil
+	}
+
+	if match.Scope == apiextensionsv1beta1.NamespaceScoped &&
+		meta.GetNamespace() == "" {
+		return false, nil
+	}
+
+	found := false
+	for _, n := range match.Namespaces {
+		if meta.GetNamespace() == n {
+			found = true
+			break
+		}
+	}
+	if !found && len(match.Namespaces) > 0 {
+		return false, nil
+	}
+
+	for _, n := range match.ExcludedNamespaces {
+		if meta.GetNamespace() == n {
+			return false, nil
+		}
+	}
+	if match.LabelSelector != nil {
```

Comment:
Should we also match the label selector against old object?

It seems weird, but that's how MutatingWebhookConfiguration's logic works:

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#mutatingwebhook-v1-admissionregistration-k8s-io

Reply:
I assumed it was only against the old object, as it's the criteria for applying a given mutation.

Given all the stability premises and the fact that we are only adding (and not changing) labels, I think it's safe to assume that if the old object matches the selector, the new one will. If only the new one will, it will be a matter of another round of mutation before the stability.

Reply:
I agree the old object/new object premise doesn't really work for mutation. It's a bit weird to be deviating from K8s convention and what we do with admission control, but I think it makes sense.

@shomron @ritazh @sozercan @brycecr thoughts?

Reply:
This is the last bit, I think apart from this discussion this is good to go.

Reply:
Adding the above people to the review, hopefully that kick-starts the discussion.

Reply:
See the notes in the discussion on the meeting minutes:

https://docs.google.com/document/d/1A1-Q-1OMw3QODs1wT6eqfLTagcGmgzAJAjJihiO3T48/edit#heading=h.5gv5xh7vxp5i

TL;DR my opinion is we not test OldObject for mutation for now, and re-visit the question before graduating from alpha, so we can make this decision based off real usage data.

Reply:
+1 on not deviating from K8s mutatingwebhookconfig logic to avoid surprises to users

> we are only adding (and not changing) labels

what if users want to match label on the old object then mutate the object to delete that label? is that a supported scenario for MVP?

Reply:
Just to make sure I understood. 
Here defining "the old" vs "the new" is a bit tricky, as we apply a bunch of mutations at the same time, and we can't just refer to the old as "before all mutations" and new as "after all mutations" because the decision of whether applying a mutation or not is based on the match logic. I think what we do here is to test only against "the old" object, with respect to the single mutation.

See https://github.com/open-policy-agent/gatekeeper/blob/624bfcc9ede7540094ea534df677479290c80f11/pkg%2Fmutation%2Fsystem.go#L85 

In that case, what is referred as old vs new? And what should happen if we find out that after a mutation the object does not match anymore?
Is it fine to assume that we match against the old, we mutate and then we "rollback" the mutation (i.e restoring a backup of the object) if match of the single mutation fails afterwards? 

Re: deleting labels, this is not possible with the current design as we are only adding non-existing labels / annotations. This is needed for making sure the mutation converges.

Reply:
Given all the reasons above, +1 on not testing OldObject for now

> Re: deleting labels, this is not possible with the current design as we are only adding non-existing labels / annotations. This is needed for making sure the mutation converges.

Thanks!

Reply:
Lemme make sure I'm getting this straight:
If we also match against newObject, then, if a mutation were to add a label to a resource, then _all_ instances of that resource type would match. The other angle on that would be that you couldn't add a label to a subset of objects of one GVK by selecting by label.

I think I don't grok the purpose of applying to both old and new objects unless you allow the mutation webhook to create or delete objects. It seems unexpected at best.

I'm in favor of shirking the k8s standard here and doing what makes more common sense, but I suspect I am missing something, too.

Reply:
One point of confusion is `OldObject` has a specific meaning, defined here:

https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/admission/v1beta1/types.go#L101-L106

I don't think that changes much, if any, of the above analysis, but has been the source of a bit of confusion.